msg.go

```diff
@@ -1057,6 +1053,86 @@ func compressionLenSearch(c map[string]int, s string) (int, bool, int) {
 	return 0, false, fullSize + len(s)
 }
 
+// Truncate removes resource records from Additional, Ns, and Answer sections to
+// ensure the message doesn't exceed the given response buffer size. It doesn't
+// alter the message if it already fits. If resource records have to be removed
+// from the Answer section, the TC header bit will be set.
```

Comment:
Required records are not confined to the Answer section... in referrals and negative responses, they appear in Additional.